crates/holochain/src/core/queue_consumer/app_validation_consumer.rs

```diff
@@ -19,6 +22,7 @@ pub fn spawn_app_validation_consumer(
 ) -> TriggerSender {
     let (tx, rx) = TriggerSender::new();
     let workspace = Arc::new(workspace);
+    let validation_dependencies = Arc::new(Mutex::new(HashSet::new()));
```

Comment:
Just so I can put it out of my mind while reading the rest of this code. The next step is planning to make this a keyed collection right? So that we know which callback needed this dependency. Because each validation callback needs to be able to proceed independently. It makes sense to de-duplicate fetches across that data structure but that's what I was expecting to see here.

Reply:
The next step is a keyed collection, but with a timestamp for each key, so that the app validation is not retriggered _if none of the deps could be fetched after an interval x_. This is to prevent indefinite re-triggering.

What you're proposing is another optimization, I believe. Keeping track of which deps belong to which op, and skip its validation if deps are missing. I can do that in a separate PR.

De-duplicating fetches is included in this PR, further down https://github.com/holochain/holochain/pull/3433/files#diff-62b92dee463214ee3a6a703cb367099a8373bdd10265616c958020c04fca75acR660-R661

Reply:
It is another optimisation and one that I think is worth making. We already did this in sys validation where it won't re-process ops that have outstanding dependencies to fetch. Or at least, that's how it's supposed to work :)

Fine with it being on another PR and what you've already done and are planning to do are all good - this is certainly something else to go alongside them